Sources/PackageLoading/ToolsVersionLoader.swift

```diff
@@ -126,58 +297,287 @@ public class ToolsVersionLoader: ToolsVersionLoaderProtocol {
+        guard let manifestContentsDecodedWithUTF8 = manifestContents.validDescription else {
+            throw Error.nonUTF8EncodedManifest(path: file)
+        }
+        
+        /// The manifest represented in its constituent parts.
+        let manifestComponents = ToolsVersionLoader.split(manifestContentsDecodedWithUTF8)
+        /// The Swift tools version specification represented in its constituent parts.
+        let toolsVersionSpecificationComponents = manifestComponents.toolsVersionSpecificationComponents
+        
+        // FIXME: Check the version specifier first?
+        // The diagnosis of the Swift tools version specification's correctness goes in the following order:
+        // 1. Check that the comment marker, the label, and the version specifier are not missing (empty).
+        // 2. Check that the everything up to the version specifier is formatted correctly according to the relaxed rules since Swift 5.3.1. Backward-compatibility is not considered here, because the user-specified version is unknown yet.
+        // 3. Check that the version spicier is formatted correctly.
+        // 4. Check that the label is formatted backward-compatibly, now that the user-specified version is known.
+        
+        let commentMarker = toolsVersionSpecificationComponents.commentMarker
+        guard !commentMarker.isEmpty else {
+            throw Error.malformedToolsVersionSpecification(.commentMarker(.isMissing))
+        }
+        
+        let label = toolsVersionSpecificationComponents.label
+        guard !label.isEmpty else {
+            throw Error.malformedToolsVersionSpecification(.label(.isMissing))
+        }
+        
+        let versionSpecifier = toolsVersionSpecificationComponents.versionSpecifier
+        guard !versionSpecifier.isEmpty else {
+            throw Error.malformedToolsVersionSpecification(.versionSpecifier(.isMissing))
+        }
+        
+        guard toolsVersionSpecificationComponents.everythingUpToVersionSpecifierIsWellFormed else {
+            if commentMarker != "//" {
+                throw Error.malformedToolsVersionSpecification(.commentMarker(.isMisspelt(String(commentMarker))))
             }
-            // Otherwise assume the default to be v3.
-            return .v3
+            
+            if label.lowercased() != "swift-tools-version:" {
+                throw Error.malformedToolsVersionSpecification(.label(.isMisspelt(String(label))))
+            }
+            
+            // The above If-statements should have covered all possible malformations in Swift tools version specification up to the version specifier.
+            // If you changed the logic in this file, and this fatal error is triggered, then you need to re-check the logic, and make sure all possible error conditions are covered in the Else-block.
+            fatalError("unidentified malformation in the Swift tools version specification")
```

Comment:
Perhaps this diagnostic should be reworded to be more approachable, especially if English isn't a first language.

Reply:
Would this be better: 

`"the Swift tools version specification has a formatting error, but the package manager is unable to diagnose it; consider replacing it with '// swift-tools-version:\(ToolsVersion.currentVersion)' to specify the current Swift toolchain version as the lowest supported version by the project; additionally, please consider filing a bug report on https://bugs.swift.org" with this file attached`

The `fatalError` is mostly aimed at SwiftPM developers, but I can see that there is a small possibility that a user might encounter an error unaccounted for by SwiftPM.

Reply:
Thanks, that makes sense.  That raises a good point actually, and one that I completely missed originally while focusing just on the text:  because `fatalError` would cause a crash rather than a diagnostic, it would be better to throw an error here with the same text.  Sorry for not spotting that earlier.

Reply:
> because `fatalError` would cause a crash rather than a diagnostic

Thanks for pointing this out! I didn't realise this either. I completely forgot that `fatalError` terminates the program right away and cannot be handled.

I changed both `fatalError`s (unidentifiable formatting error in the Swift tools version specification and unidentifiable backward-incompatibility in the manifest) to some forms of `ToolsVersionLoader.Error` in c1e5340fcf6952d8bed7355114b2f40befcc885c, and reworded the error messages. The messages are now more detailed and more user-friendly, with suggestions for both potential recovery and bug-reporting.

Reply:
Thanks for making this change!  As these errors are not supposed to be hit normally, I think the test coverage you added is enough.